flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java

```diff
@@ -257,6 +257,18 @@
 					text("\"ip\" - uses host's ip address as binding address"))
 				.build());
 
+	/**
+	 * The prefix of the TaskManager's ResourceID. If configured, the ResourceID will be generated with the prefix and a
+	 * 6-character random string. Otherwise, a random string will be used. Only valid in standalone mode.
+	 */
+	@Documentation.Section(Documentation.Sections.ALL_TASK_MANAGER)
+	public static final ConfigOption<String> TASK_MANAGER_RESOURCE_ID_PREFIX =
+		key("taskmanager.resource-id.prefix")
+			.stringType()
+			.noDefaultValue()
+			.withDescription("The prefix of the TaskManager's ResourceID. If configured, the ResourceID will be generated with the prefix "
+				+ "and a 6-character random string. Otherwise, a random string will be used. Only valid in standalone mode.");
```

Comment:
Why is this only valid in standalone mode? Shouldn't we have a general mean how to configure the resource id for `TaskExecutors` in general? I believe it should work across the different integrations (Yarn, Mesos, K8s, standalone) all the same. The only difference would be to what value the different implementations set it.

Reply:
Thanks for raising this valid concern. For standalone cluster, we could set different `TASK_MANAGER_RESOURCE_ID_PREFIX` per `TaskExecutor`. Currently, for active resourcemanager(Yarn, Mesos, K8s), we will generate a meaningful `ResourceID`(e.g. container id in Yarn, pod name in native K8s) per `TaskExecutor`. If we also want to respect the `TASK_MANAGER_RESOURCE_ID_PREFIX` config option, then it will be difficult to use the `ResourceID` to find the corresponding container/pod.

Given that `TASK_MANAGER_RESOURCE_ID_PREFIX` is configured to "my-yarn-cluster-tm", then all the `TaskExecutor` share the same schema "my-yarn-cluster-tm-xxxxxx". A corresponding container id could be "container_e01_1592277741400_0043_01_000002". It is really confusing.

If you just mean to override `TASK_MANAGER_RESOURCE_ID_PREFIX` option for active deployment with current meaningful name, i think it makes sense. We just use the config option to replace the environment to pass the `ResourceID` to `TaskExecutor`.

Reply:
Yes what I mean is to have a single mean how the `ResourceID` is configured for across the different integrations (Yarn, Mesos, K8s). Maybe we should take a step back and not start with a prefix config option but simply with `TASK_MANAGER_RESOURCE_ID` which allows to set the `ResourceID`. That way it would be simple for the RM implementations to configure the `ResourceID`. Per default it is empty and then we could take the hostname and a random value, for example.

Reply:
Thanks for the discussion. Add environment variable `TASK_MANAGER_RESOURCE_ID` 
 also sounds good. However, as @azagrebin said in FLINK-17579
> there are plans in the community to introduce a unified approach to set Flink options via environment variables.

Not sure is it ok to apply this approach. Should we move the discussion back to FLINK-17579?

Reply:
Yes, actually in FLINK-17579 we also want to start with environment instead of config option. And we could unify all the current environment name(e.g. `_FLINK_CONTAINER_ID`, `_FLINK_POD_NAME`) to `TASK_MANAGER_RESOURCE_ID`. For active resourcemanager, we generate with meaningful names and do not change the current behavior. And for standalone, it could be specified with an arbitrary string(e.g. hostname, podname, etc.).

Reply:
Why do you want to do it with an environment variable? Can't it be a configuration value? That way the user could also configure it simply in his `flink-conf.yaml`.

This would be a configuration value which is specific to the started process and which should not distributed to other processes, though. I think we already have similar cases in Flink. See `DeploymentOptionsInternal.CONF_DIR`, for example.

Reply:
Ahh, I think I misunderstand what you mean by "step back". You mean we could set the `ResourceID` to `host-uuid` by default. But for the RM implementations, they could override it to `containerID` through a configuration option `TASK_MANAGER_RESOURCE_ID `. Do I understand it correctly?

If that is the case, user should not set `TASK_MANAGER_RESOURCE_ID` in active mode (active RMs will override this option anyway). And in standalone mode, it is user's responsibility to ensure that these ids do not conflict.

Reply:
Yes, this was my idea. The down side would be that people could shoot themselves in the foot with it. But I believe that this should be fine given that it is an expert feature.